File: problem_003.c
// -*- compile-command: "gcc -std=c99 -o problem_003 problem_003.c  ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 3:\n"
           "What is the largest prime factor of the number 600851475143?\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    long long bigNum = 600851475143;
    int divisor = 2;

    while ( divisor < bigNum )
        if ( bigNum % divisor == 0 )
            bigNum /= divisor;
        else
            divisor++;

    printf("Largest factor found: %d\n", divisor);
    printElapsedTime(start);

    return 0;
}

File: problem_003.rs
// -*- compile-command: "rustc -o problem_003_rs problem_003.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 {
        return true;
    }
    if n % 2 == 0 {
        return false;
    }

    let limit = (n as f64).sqrt() as u64;
    for i in 3..(limit+1) {
        if n % i == 0 {
            return false;
        }
    }
    return true;
}


fn main() {
    println!("Project Euler - Problem 3");
    println!("What is the largest prime factor of the number 600851475143?\n");

    let mut n: u64 = 600851475143;
    let mut d: u64 = 3;
    while n > 1 {
        d += 2;
        while is_prime(d) == false {
            d += 2;
        }

        while n % d == 0 {
            n = n / d;
        }
    }

    println!("{}", d);
}
